lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+  using ParseKind = parser::InquireSpec::IntVar::Kind;
+  switch (std::get<parser::InquireSpec::IntVar::Kind>(spec.t)) {
+  case ParseKind::Iostat: specKind = SpecifierKind::Iostat; break;
+  case ParseKind::Nextrec: specKind = SpecifierKind::Nextrec; break;
+  case ParseKind::Number: specKind = SpecifierKind::Number; break;
+  case ParseKind::Pos: specKind = SpecifierKind::Pos; break;
+  case ParseKind::Recl: specKind = SpecifierKind::Recl; break;
+  case ParseKind::Size: specKind = SpecifierKind::Size; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::InquireSpec::LogVar &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::InquireSpec::LogVar::Kind;
+  switch (std::get<parser::InquireSpec::LogVar::Kind>(spec.t)) {
+  case ParseKind::Exist: specKind = SpecifierKind::Exist; break;
+  case ParseKind::Named: specKind = SpecifierKind::Named; break;
+  case ParseKind::Opened: specKind = SpecifierKind::Opened; break;
+  case ParseKind::Pending: specKind = SpecifierKind::Pending; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec &spec) {
+  // IoControlSpec context Name
+  hasIoControlList_ = true;
+  if (const parser::Name * name{std::get_if<parser::Name>(&spec.u)}) {
+    CHECK(name->symbol != nullptr && name->symbol->has<NamelistDetails>());
+    SetSpecifier(SpecifierKind::Nml);
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Asynchronous &spec) {
+  SetSpecifier(SpecifierKind::Asynchronous);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      hasAsynchronousYes_ = parser::ToUpperCaseLetters(**charConst) == "YES"s;
+      CheckStringValue(SpecifierKind::Asynchronous, **charConst,
+          parser::FindSourceLocation(spec));  // C1223
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::CharExpr &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::IoControlSpec::CharExpr::Kind;
+  switch (std::get<ParseKind>(spec.t)) {
+  case ParseKind::Advance: specKind = SpecifierKind::Advance; break;
+  case ParseKind::Blank: specKind = SpecifierKind::Blank; break;
+  case ParseKind::Decimal: specKind = SpecifierKind::Decimal; break;
+  case ParseKind::Delim: specKind = SpecifierKind::Delim; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  }
+  SetSpecifier(specKind);
+  if (const auto *expr{GetExpr(std::get<parser::ScalarDefaultCharExpr>(spec.t)
+                                   .thing.thing.value())}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
```

Comment:
It's okay to use `auto` here rather than redundantly repeating the type, if you like (and fine if you don't).

Reply:
Done.  (The single remaining instance of this code is now in check-io.h.)